# Table of Contents
 <p><div class="lev1 toc-item"><a href="#1.-Data-exploration" data-toc-modified-id="1.-Data-exploration-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>1. Data exploration</a></div><div class="lev2 toc-item"><a href="#Load-in-the-schema-of-the-data." data-toc-modified-id="Load-in-the-schema-of-the-data.-11"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Load in the schema of the data.</a></div><div class="lev2 toc-item"><a href="#Import-sample-of-the-data" data-toc-modified-id="Import-sample-of-the-data-12"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Import sample of the data</a></div><div class="lev2 toc-item"><a href="#Original-Data-stats" data-toc-modified-id="Original-Data-stats-13"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Original Data stats</a></div><div class="lev3 toc-item"><a href="#Total-Number-of-Tweets" data-toc-modified-id="Total-Number-of-Tweets-131"><span class="toc-item-num">1.3.1&nbsp;&nbsp;</span>Total Number of Tweets</a></div><div class="lev3 toc-item"><a href="#Column-formats" data-toc-modified-id="Column-formats-132"><span class="toc-item-num">1.3.2&nbsp;&nbsp;</span>Column formats</a></div><div class="lev3 toc-item"><a href="#Column-Descriptions" data-toc-modified-id="Column-Descriptions-133"><span class="toc-item-num">1.3.3&nbsp;&nbsp;</span>Column Descriptions</a></div><div class="lev2 toc-item"><a href="#Data-Preprocessing" data-toc-modified-id="Data-Preprocessing-14"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Data Preprocessing</a></div><div class="lev3 toc-item"><a href="#Remove-unnecessary-columns" data-toc-modified-id="Remove-unnecessary-columns-141"><span class="toc-item-num">1.4.1&nbsp;&nbsp;</span>Remove unnecessary columns</a></div><div class="lev3 toc-item"><a href="#NaN-Cells" data-toc-modified-id="NaN-Cells-142"><span class="toc-item-num">1.4.2&nbsp;&nbsp;</span>NaN Cells</a></div><div class="lev3 toc-item"><a href="#Convert-'createdAt'-to-datetime-type" data-toc-modified-id="Convert-'createdAt'-to-datetime-type-143"><span class="toc-item-num">1.4.3&nbsp;&nbsp;</span>Convert 'createdAt' to datetime type</a></div><div class="lev4 toc-item"><a href="#Observation" data-toc-modified-id="Observation-1431"><span class="toc-item-num">1.4.3.1&nbsp;&nbsp;</span><em>Observation</em></a></div><div class="lev3 toc-item"><a href="#Remove-and-Replace-Urls/UserName/Hashtags-from-tweets" data-toc-modified-id="Remove-and-Replace-Urls/UserName/Hashtags-from-tweets-144"><span class="toc-item-num">1.4.4&nbsp;&nbsp;</span>Remove and Replace Urls/UserName/Hashtags from tweets</a></div><div class="lev4 toc-item"><a href="#Observation" data-toc-modified-id="Observation-1441"><span class="toc-item-num">1.4.4.1&nbsp;&nbsp;</span><em>Observation</em></a></div><div class="lev2 toc-item"><a href="#Data-Stats" data-toc-modified-id="Data-Stats-15"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Data Stats</a></div><div class="lev3 toc-item"><a href="#Language-Usage" data-toc-modified-id="Language-Usage-151"><span class="toc-item-num">1.5.1&nbsp;&nbsp;</span>Language Usage</a></div><div class="lev4 toc-item"><a href="#Observation" data-toc-modified-id="Observation-1511"><span class="toc-item-num">1.5.1.1&nbsp;&nbsp;</span><em>Observation</em></a></div><div class="lev3 toc-item"><a href="#Number-of-Tweets-per-User" data-toc-modified-id="Number-of-Tweets-per-User-152"><span class="toc-item-num">1.5.2&nbsp;&nbsp;</span>Number of Tweets per User</a></div><div class="lev4 toc-item"><a href="#Observation" data-toc-modified-id="Observation-1521"><span class="toc-item-num">1.5.2.1&nbsp;&nbsp;</span><em>Observation</em></a></div><div class="lev3 toc-item"><a href="#Number-of-Users-in-Sample-Data" data-toc-modified-id="Number-of-Users-in-Sample-Data-153"><span class="toc-item-num">1.5.3&nbsp;&nbsp;</span>Number of Users in Sample Data</a></div><div class="lev3 toc-item"><a href="#Percentage-with-hashtags" data-toc-modified-id="Percentage-with-hashtags-154"><span class="toc-item-num">1.5.4&nbsp;&nbsp;</span>Percentage with hashtags</a></div><div class="lev4 toc-item"><a href="#Number-of-tweets-per-hashtag" data-toc-modified-id="Number-of-tweets-per-hashtag-1541"><span class="toc-item-num">1.5.4.1&nbsp;&nbsp;</span>Number of tweets per hashtag</a></div><div class="lev4 toc-item"><a href="#Usage-of-shared-/-common-tags" data-toc-modified-id="Usage-of-shared-/-common-tags-1542"><span class="toc-item-num">1.5.4.2&nbsp;&nbsp;</span>Usage of shared / common tags</a></div><div class="lev4 toc-item"><a href="#Observation" data-toc-modified-id="Observation-1543"><span class="toc-item-num">1.5.4.3&nbsp;&nbsp;</span><em>Observation</em></a></div><div class="lev3 toc-item"><a href="#0.3.1-Percentage-with-geolocation" data-toc-modified-id="0.3.1-Percentage-with-geolocation-155"><span class="toc-item-num">1.5.5&nbsp;&nbsp;</span>0.3.1 Percentage with geolocation</a></div><div class="lev2 toc-item"><a href="#Emoji" data-toc-modified-id="Emoji-16"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Emoji</a></div>

> **Purpose of this notebook: **

> Before diving to the huge datasets provided by Swisscom, need to understand it and understand what we want to do with it. So we will use this notebook with a small sample of the datasets, to simulate the steps of our project. The final goal is to create a python scipt, so we just need to launch it on the cluster to obtain the desired results.

**Import Packages**

In [245]:
import pandas as pd
import numpy as np
import csv

# plots
from matplotlib.pyplot import *
matplotlib.style.use('ggplot')

from pylab import *

from pandas import Series
import re

# language detection
from langdetect import detect

In [246]:
print("Pandas version {}".format(pd.__version__))

Pandas version 0.19.2


In [247]:
#increase the max column width displayed from the default 50
pd.set_option('display.max_colwidth', 200)

# 1. Data exploration

***
## Load in the schema of the data.


Note that *schema.txt* file given in the epfl cluster was manually modified in order to import it easily with pandas.

In [303]:
# read schema 
schema_rawfile = pd.read_csv("twitter-swisscom/schema_home.txt", header=None, sep='\s+')
# extract schema columns
twitter_data_columns = schema_rawfile[1].values

print("Nbr features:",len(twitter_data_columns))
list(twitter_data_columns)

Nbr features: 20


['id',
 'userId',
 'createdAt',
 'text',
 'longitude',
 'latitude',
 'placeId',
 'inReplyTo',
 'source',
 'truncated',
 'placeLatitude',
 'placeLongitude',
 'sourceName',
 'sourceUrl',
 'userName',
 'screenName',
 'followersCount',
 'friendsCount',
 'statusesCount',
 'userLocation']

***
## Import sample of the data


We will not work directly with the entire datasets. Instead, we prepare our code and function here, with a small sample of the data. Then we create a script to be executed in the EPFL cluster.


**First step:** "Clean" the tsv file. The files we are given are in a *.tsv* format: there is one tweet per line and each features within a tweet are separated by a tab. Unfortunately, some tweets text contain break line (*\n*). If we try to import them like that, python will take those break line *\n* as a new line, therefore as a new tweet. 

But the break line inside the tweets are preceed by a backslash. Therefore, we run a bash command to replace all combination of *\ + \n* by a simple white space.

>     sed -e :a -e '/\\$/N; s/\\\n/ /; ta' twitter-swisscom/sample.tsv  > data_clean/sample.tsv

**Second step**: The file *data_clean/sample.tsv* has the following attributes:

    *application/octet-stream; charset=binary*

Therefore, we cannot load it *as is* into a pandas Dataframe. We will convert it to a simple *text/plain* with encoding *utf-8*, that pandas can perfectly handle.

>     cat data_clean/sample.tsv | tr -d '\0' > data_clean/sample_formatted.tsv

**Third step**: Now we can import the data to a pandas dataframe.

In [249]:
sample_data = pd.read_table(r"data_clean/sample_formatted.tsv",sep='\t', quoting=csv.QUOTE_NONE, header=None, names=twitter_data_columns)
sample_data.head()

id              userId            createdAt  \
0  776522983837954049  735449229028675584  2016-09-15 20:48:01   
1  776523000636203010          2741685639  2016-09-15 20:48:05   
2  776523045200691200           435239151  2016-09-15 20:48:15   
3  776523058404290560           503244217  2016-09-15 20:48:18   
4  776523058504925185           452805259  2016-09-15 20:48:18   

                                                                                                                                     text  \
0                                                                                                se lo dici tu... https://t.co/x7Qm1VHBKL   
1                                                                                                                 https://t.co/noYrTnqmg9   
2  @BesacTof @Leonid_CCCP Tu dois t'engager en signant précisément:"Je partage les valeurs républicaines de la droite et du centre 1/2...   
3    @Mno0or_Abyat اشوف مظاهرات على قانون العمل الجديد انا طلعت الصباح من هناك مافيه شي بس اشوف الأخبار قبل شوي المشكلة راجع بعد يومين !!   
4                                                                       Greek night #geneve (@ Emilios in Genève) https://t.co/sEplW0Mcyz   

  longitude latitude           placeId           inReplyTo  source  truncated  \
0        \N       \N  51c0e6b24c64e54e                  \N       1        NaN   
1        \N       \N  4e7c21fd2af027c6                  \N       1        NaN   
2        \N       \N  12eb9b254faf37a3  776522113859608576       5        NaN   
3        \N       \N  30bcd7f767b4041e  776521597515624448       1        NaN   
4   6.14414  46.1966  c3a6437e1b1a726d                  \N       3        NaN   

   placeLatitude  placeLongitude           sourceName  \
0        46.0027         8.96044   Twitter for iPhone   
1        46.8131         8.22414   Twitter for iPhone   
2        47.2010         5.94082  Twitter for Android   
3        45.8011         6.16552   Twitter for iPhone   
4        46.2048         6.14319           foursquare   

                               sourceUrl         userName screenName  \
0  http://twitter.com/#!/download/iphone  plvtone filiae.  hazel_chb   
1  http://twitter.com/#!/download/iphone           samara   letisieg   
2    http://twitter.com/download/android        lebrübrü❤   lebrubru   
3  http://twitter.com/#!/download/iphone   عبدالله القنيص   bingnais   
4                  http://foursquare.com      Alkan Şenli   Alkanoli   

   followersCount  friendsCount  statusesCount        userLocation  
0             146           110          28621           Earleen.   
1             755          2037           3771              Suisse  
2             811           595          30191             Fontain  
3           28433           417          12262           Shargeyah  
4             204           172           3390  İstanbul/Burgazada

***
## Original Data stats


### Total Number of Tweets

We have **_8790_** entries in our dataframe and each row contains **_20_** columns.

In [250]:
TOTAL_ENTRIES = len(sample_data)
TOTAL_COLUMNS = len(sample_data.columns)

print(TOTAL_ENTRIES)
print(TOTAL_COLUMNS)

8790
20


### Column formats

In [251]:
sample_data.dtypes

id                  int64
userId              int64
createdAt          object
text               object
longitude          object
latitude           object
placeId            object
inReplyTo          object
source              int64
truncated         float64
placeLatitude     float64
placeLongitude    float64
sourceName         object
sourceUrl          object
userName           object
screenName         object
followersCount      int64
friendsCount        int64
statusesCount       int64
userLocation       object
dtype: object

### Column Descriptions
We've gathered more detailed dscription of each attribute from the [Twtter API](https://dev.twitter.com/overview/api/users)

In [252]:
data_description = pd.read_csv("twitter-swisscom/schema_description.txt", header=None, names=['Attribute', 'Description'], sep='\s\s+', engine='python', index_col=0)
data_description

Attribute  \
1               id   
2           userId   
3        createdAt   
4             text   
5        longitude   
6         latitude   
7          placeId   
8        inReplyTo   
9           source   
10       truncated   
11   placeLatitude   
12  placeLongitude   
13      sourceName   
14       sourceUrl   
15        userName   
16      screenName   
17  followersCount   
18    friendsCount   
19   statusesCount   
20    userLocation   

                                                                                                                                                                                                Description  
1                                                                                                                                       The integer representation of the unique identifier for this Tweet.  
2                                                                                                                                        The integer representation of the unique identifier for this User.  
3                                                                                                              UTC time when this Tweet was created. Example: "created_at":"Wed Aug 27 13:08:45 +0000 2008"  
4                                                                                                                                                               The actual UTF-8 text of the status update.  
5                                                                                                  Nullable Represents the geographic location of this Tweet as reported by the user or client application.  
6                                                                                                  Nullable Represents the geographic location of this Tweet as reported by the user or client application.  
7   ID representing this place. Places are specific, named locations with corresponding geo coordinates. They can be attached to Tweets by specifying a place_id when tweeting. Tweets associated with p...  
8                                                                              Nullable If the represented Tweet is a reply, this field will contain the integer representation of the original Tweet’s ID.  
9                                                                                                                                                                                                      None  
10  Indicates whether the value of the text parameter was truncated, for example, as a result of a retweet exceeding the 140 character Tweet length. Truncated text will end in ellipsis, like this ... ...  
11                                                                                                                                                                                              see placeID  
12                                                                                                                                                                                              see placeID  
13                                                                                                                                                               Name of the utility used to post the Tweet  
14                                                                                                                                                                Url of the utility used to post the Tweet  
15                                                                  The name of the user, as they’ve defined it. Not necessarily a person’s name. Typically capped at 20 characters, but subject to change.  
16  The screen name, handle, or alias that this user identifies themselves with. screen_names are unique but subject to change. Typically a maximum of 15 characters long, but some historical accounts ...  
17                                          

***
## Data Preprocessing

Before analysing the sentiment of tweets and extract it's location information, we have to pre-process to prepare our data.

### Remove unnecessary columns


In [253]:
sample_data = sample_data.drop('source', 1)
sample_data = sample_data.drop('truncated', 1)
sample_data = sample_data.drop('sourceName', 1)
sample_data = sample_data.drop('sourceUrl', 1)
sample_data = sample_data.drop('followersCount', 1)
sample_data = sample_data.drop('friendsCount', 1)
sample_data = sample_data.drop('statusesCount', 1)
sample_data = sample_data.drop('userName', 1)

In [254]:
list(sample_data.columns)

['id',
 'userId',
 'createdAt',
 'text',
 'longitude',
 'latitude',
 'placeId',
 'inReplyTo',
 'placeLatitude',
 'placeLongitude',
 'screenName',
 'userLocation']

### NaN Cells

The **text** column (tweet content) contains NaN Value. We have no choice but to remove them because without the content we won't be able to proceed and perform sentiment analysis. As for **truncated** and **userName**, we don't worry much as truncated is just a flag indicating whether the content is above 140 characters and userId is better and more unique than userName.

***Note: ***missing data are represented as **"\N"** in the dataframe.

In [255]:
sample_data.isnull().any()

id                False
userId            False
createdAt         False
text               True
longitude         False
latitude          False
placeId           False
inReplyTo         False
placeLatitude     False
placeLongitude    False
screenName        False
userLocation      False
dtype: bool

In [256]:
# Create a true/false list then map to the original dataframe
NaN_text = sample_data.text.isnull()
sample_data = sample_data[~ NaN_text]

In [257]:
print("Number of entries missing content - ", NaN_text.sum())
print("Entries in sample_data dataframe - ", sample_data.text.count())
print("Percentage of missing data", 93/8790)

Number of entries missing content -  93
Entries in sample_data dataframe -  8697
Percentage of missing data 0.010580204778156996


### Convert 'createdAt' to datetime type


The **createdAt** column is of type String and contains the UTC time when this Tweet was created. However we want to convert this to the datatime Python type for the ease of manipulation in future analysis.

In [258]:
sample_data.dtypes[:3]

id            int64
userId        int64
createdAt    object
dtype: object

In [259]:
# Convert the type
sample_data['createdAt'] = pd.to_datetime(sample_data['createdAt'])
sample_data.dtypes[:3]

id                    int64
userId                int64
createdAt    datetime64[ns]
dtype: object

In [260]:
# We can now sort our dataframe based on createdAt column
sample_data = sample_data.sort_values('createdAt', axis=0, ascending=False)

In [261]:
sample_data.createdAt.describe()

count                    8697
unique                   8070
top       2016-09-16 08:25:00
freq                        4
first     2016-09-15 20:48:01
last      2016-09-16 16:35:19
Name: createdAt, dtype: object

#### *Observation*

There are *8070* unique datatimes in our dataframe, and all of our tweets are created between *2016-09-15 20:48:01* and *2016-09-16 08:25:00*

### Remove and Replace Urls/UserName/Hashtags from tweets


We want to remove urls and usernames from the content of the tweet because it gives us no information about the sentiment and the package we use to detect what language the tweet is written can't process a single url as the text has no feature to analyse.

***Note: ***The original idea was to replace url with "URL" and @user to "AT_USER", however a single tweet contains either or both URL link and @user have no feature to provide to sentiment analysis and language detection, thus we have decided to replace it with empty String and remove the entire entry is there's nothing left after replacement.

In [262]:
sample_data['hashtags'] = ""

for index, row in sample_data.iterrows():
    # match url in string and replace with "" empty string
    t = re.sub(r"http\S+", "", row['text'])
    # remove @username
    t = re.sub('@[^\s]+', "", t)
    # find all hashtags
    tag_list = re.findall(r'#(\w+)', t)
    # remove hashtags
    t = re.sub(r'#(\w+)', "",  t)
    # replace multiple whitespaces to one
    t = re.sub('[\s]+', " ", t)
    # trim
    t = t.strip()
    # fill in the hashtag column is the list is not empty
    if (tag_list):
        sample_data.set_value(index, "hashtags", " ".join(tag_list))
    if (t== ""): # remove the row if the result is empty after stripping
        sample_data = sample_data.drop(index)
    else:
        sample_data.set_value(index, "text", t)


In [263]:
count = sample_data.text.count()

print("There are {} entries in the dataframe".format(count))

There are 8308 entries in the dataframe


#### *Observation*

From our inspection, most urls and usernames are removed from the tweet. As a result we have shrunk the dataset to {{count}} entries which is {{float("{0:.2f}".format(count/TOTAL_ENTRIES * 100))}}% of the original dataset 

**Note**: Although Urls and Usernames don't contain any feature for sentiment analysis, but it could be the link to connect different tweets together (different user discussing/expressing their thought about the same topic).

***
## Data Stats

### Language Usage


We use [langdetect](https://pypi.python.org/pypi/langdetect) python package to determine what language is used for the tweet. Of course come tweets are written in multiple languages, this library returns the most probable language after analysis.

In [264]:
TOTAL_ENTRIES = sample_data.text.count()

In [265]:
sample_data['language'] = ''

for index, row in sample_data.iterrows():
    lang = "None"
    try:
        lang = detect(row['text'])
    except:
        pass
    sample_data.set_value(index, "language", lang)

In [266]:
def d(x):
    return Series(dict(Number_of_tweets = x['text'].count(),
                       Percentage = x['text'].count() / TOTAL_ENTRIES * 100,
                       ))

In [267]:
lang_usage = sample_data.groupby('language').apply(d).sort_values('Percentage', axis=0, ascending=False)

lang_usage

Number_of_tweets  Percentage
language                              
en                  2140.0   25.758305
fr                  1689.0   20.329803
de                  1208.0   14.540202
it                   718.0    8.642273
ar                   393.0    4.730380
es                   307.0    3.695234
pt                   252.0    3.033221
tr                   169.0    2.034184
so                   117.0    1.408281
tl                   108.0    1.299952
nl                   108.0    1.299952
id                   100.0    1.203659
no                    94.0    1.131440
None                  90.0    1.083293
af                    84.0    1.011074
ja                    79.0    0.950891
ru                    69.0    0.830525
ca                    66.0    0.794415
da                    50.0    0.601830
fi                    45.0    0.541647
sv                    43.0    0.517573
ro                    40.0    0.481464
cy                    40.0    0.481464
et                    38.0    0.457390
vi                    36.0    0.433317
pl                    28.0    0.337025
sw                    24.0    0.288878
fa                    22.0    0.264805
hr                    22.0    0.264805
hu                    20.0    0.240732
sl                    20.0    0.240732
th                    16.0    0.192585
lv                    15.0    0.180549
sq                    13.0    0.156476
lt                    10.0    0.120366
mk                     8.0    0.096293
sk                     8.0    0.096293
cs                     6.0    0.072220
bg                     5.0    0.060183
uk                     3.0    0.036110
el                     2.0    0.024073
he                     1.0    0.012037
ur                     1.0    0.012037
ko                     1.0    0.012037

#### *Observation*

**English** is the most used language written ***25.9%*** of total tweets followed by ***20.3%*** of **French**, ***14.4%*** of **German**, ***8.6%*** of **Italian**, ***4.7%*** of **Arabic** and ***3.6%*** of **Spanish**. 

[langdetect](https://pypi.python.org/pypi/langdetect) wasn't able to determine the language used for 1.08% percentage of the tweets.

### Number of Tweets per User 


In [268]:
def f(x):
    return Series(dict(Number_of_tweets = x['text'].count(),
                       tweet_id = x['id'].min(),
                       ))

In [269]:
tweet_count = sample_data.groupby('userId').apply(f)
tweet_count = tweet_count.sort_values('Number_of_tweets', axis=0, ascending=False)
tweet_count.head()

Number_of_tweets            tweet_id
userId                                                  
766674630014857217               217  776525594515349505
758247740019126272               185  776550140828119040
735449229028675584               111  776522983837954049
570608126                         93  776664047815106560
866661589                         87  776525846563659776

In [270]:
tweet_count.Number_of_tweets.describe()

count    2651.000000
mean        3.133912
std         8.303113
min         1.000000
25%         1.000000
50%         1.000000
75%         3.000000
max       217.000000
Name: Number_of_tweets, dtype: float64

#### *Observation*

More than 50% of the user only had 1 tweet in our sample data, total avg is 3 tweets per user.

***
### Number of Users in Sample Data


All **{{TOTAL_ENTRIES}}** entries in our dataframe have an unique id. However there are **2651** unique users in our data. 

In [271]:
total_tweets = len(pd.unique(sample_data.id.ravel()))
total_unique_user = len(pd.unique(sample_data.userId.ravel()))

print("Percentage of {}(users) / {}(tweets) = {}%".format(total_unique_user, total_tweets, total_unique_user/total_tweets))

Percentage of 2651(users) / 8308(tweets) = 0.3190900337024555%


***
### Percentage with hashtags

We want to know the amount of data contain hashtag info. This could give us a hint whether hashtags are used significantly by users and could potential give us location/event information for future sentiment analysis.

#### Number of tweets per hashtag

***Note: ***We did not split the entry which contain multiple hashtags because the result from a simple groupby generates what we hoped to obtain.

In [297]:
def f(x):
    return Series(dict(Number_of_tweets = x['text'].count(),
                       Percentage = x['text'].count() / TOTAL_ENTRIES * 100,
                       ))

In [298]:
hashtag_count = sample_data.groupby('hashtags').apply(f)
hashtag_count = hashtag_count.sort_values('Number_of_tweets', axis=0, ascending=False)
number_without_tag = hashtag_count.Number_of_tweets.get_value(0)
hashtag_count.head(5)

Number_of_tweets  Percentage
hashtags                                                              
                                                    6358.0   76.528647
BFYConcertFinale                                     191.0    2.298989
DiFe16                                                27.0    0.324988
motormaniaci webtv sbandy moto instamoto              23.0    0.276842
Controleurs                                           20.0    0.240732

#### Usage of shared / common tags

**Question:** What percentage of tweets had 1/2/3.. shared tags. 

*It's a bit obscure but we want to know how many tags were shared by tweets in our dataset*

In [299]:
def f(x):
    return Series(dict(Percentage = x['Number_of_tweets'].sum() / TOTAL_ENTRIES * 100,
                       ))

In [302]:
common_tag = hashtag_count.groupby('Number_of_tweets').apply(f)
common_tag = common_tag.sort_values('Percentage', axis=0, ascending=False)
common_tag

Percentage
Number_of_tweets            
6358.0             76.528647
1.0                14.973519
191.0               2.298989
2.0                 1.949928
3.0                 0.722195
4.0                 0.481464
7.0                 0.337025
27.0                0.324988
13.0                0.312951
5.0                 0.300915
23.0                0.276842
20.0                0.240732
18.0                0.216659
6.0                 0.216659
15.0                0.180549
14.0                0.168512
12.0                0.144439
10.0                0.120366
9.0                 0.108329
8.0                 0.096293

#### *Observation*

The majority of the tweets ({{number_without_tag}} out of {{TOTAL_ENTRIES}}) have no hashtags, which is **{{float("{0:.2f}".format(number_without_tag/TOTAL_ENTRIES * 100))}}**% of the dataset. More than 50% of the rest of the dataset don't share any common hashtags.

### 0.3.1 Percentage with geolocation

In [ ]:
long = sample_data.longitude
long_not_null = (long != r'\N').sum()
print("Percentage of data with longitude info:", 100*(long_not_null/len(sample_data)))
print("Nbr: ", long_not_null)

In [ ]:
lat = sample_data.latitude
lat_not_null = (lat != r'\N').sum()
print("Percentage of data with latitude info:", 100*(lat_not_null/len(sample_data)))
print("Nbr: ", lat_not_null)

In [ ]:
long_not_null = long!=r'\N'
lat_not_null = lat!=r'\N'
diff = (long_not_null != lat_not_null).any()
print("Is there null long/lat with valid lat/long ? :", diff)

That's for the latitude/longitude... what ? The user ? 
Because we also have the fields ***placeLongitude*** and ***placeLatitude***.

From the twitter api:
> Places are specific, named locations with corresponding geo coordinates. They can be attached to Tweets by specifying a place_id when tweeting. Tweets associated with places are not necessarily issued from that location but could also potentially be about that location.

Since we want to map the sentiment of tweet with a given location, we will use fields ***placeLongitude*** and ***placeLatitude*** for this project.

In [ ]:
xx = sample_data.groupby(['placeId']).size()
xx.sort(ascending=False)
xx

***
## Emoji
